### ISSUE TITLE
Do we have a datatype like Querry

### ISSUE THREAD

**Issue by Shaunmax** [distriqt/ANE-Firebase#213]
import com.distriqt.extension.firebase.database.Query;

private var _query:Query;

##### Expected behaviour
Query datatype should show up if available.

##### Actual behaviour

No Query datatype, instead I can see Flox's Query datatype 

##### Environment 

- AIR SDK version: 
- Device OS: 
- Device version: 
- Development IDE: 
- Development OS:

##### Logs

Device logs if relevant

**Comment by marchbold** [comment]
Hi,

Not exactly sure what you are asking but here is the reference for the Database `Query` object:

https://distriqt.github.io/ANE-Firebase/asdocs/com/distriqt/extension/firebase/database/Query.html

**Comment by Shaunmax** [comment]
When I try to add a query variable the Query datatype doesnt show up.

**Comment by Shaunmax** [comment]
Here is a screenshot :-

https://gyazo.com/6d2f157f0816823ba3688bbe730f045f

**Comment by marchbold** [comment]
Not really sure why your auto complete isn't picking it up. What IDE is that? 

Just make sure you add the correct import and it should work without an issue.

```as3
import com.distriqt.extension.firebase.database.Query;
```

**Comment by Shaunmax** [comment]
adobe animate cc

**Comment by Shaunmax** [comment]
My concern is everytime i query, do I have to add new event listener??

```
public function queryYesterdaysResult(): void {

try {
var yesterdayQuery:Query = FirebaseDatabase.service.getReference("main_database").child("leaderboard").child(yesterday_server).orderByChild("gScore").limitToFirst(10);
yesterdayQuery.addEventListener(DatabaseReferenceChildEvent.CHILD_ADDED, handleYesterdayMVP);

} catch (e: Error) {
Flox.logWarning("Something went wrong in  queryYesterdaysResult!" + e.message);
}
} 
```

**Comment by marchbold** [comment]
Can you clarify what you are asking?

**Comment by Shaunmax** [comment]
If you look at the above function, for each queryYesterdaysResult(), I am adding an event listener. 
We need to do that? If I call  queryYesterdaysResult() 10 times, will it add 10 event listener. 
Do I need to use yesterdayQuery.removeEventListener(DatabaseReferenceChildEvent.CHILD_ADDED, handleYesterdayMVP); ? Whats your suggestion for code optimisation.?

**Comment by marchbold** [comment]
I would highly suggest you create a class variable and use that rather than recreating an object every time you access that function. 

Generally would never suggest implementing any code like the above as if you don't hold onto a reference of a single object you are creating multiple instances of an object and potentially leaking them as you probably never clean up that event listener. Additionally AIR may decide to garbage collect these objects and you'll run into random issues with events. 

Does that answer your question?

**Comment by Shaunmax** [comment]
That's exactly what I am trying to do, and last time when I checked, I
couldn't find the Query datatype, not sure why and whether its a IDE
problem or something else.

I will definitely try that out later as I am currently working on a
different priority.

My initial understanding was that, each time I query, it returns the result
in an array but thats not how it works. For eg, when I query for first 10,
the event listener gets triggered 10 times and I have to store each result
in an array. This is where I started facing the issue and implemented the
way it was done in one of the examples files.

But I have a doubt on memory leakage and wanted to implement it in a
different way. Thats why I was enquiring about it.

Also when request for first 10 results and if there are no 10 results, lets
say only 5 scores are there. Adding a remove listener would be tricky.
Is there a way to know the end of query result. In that case I only need to
check that and remove it when end of query result is caught.

Let me know!

**Comment by marchbold** [comment]
Yes the `CHILD_ADDED` event will be triggered multiple times as each reference is added to the query. 

If you want to just read the data once use the `once` function.

Might be worth checking the documentation so you can determine the best method to access the data you need:

https://github.com/distriqt/ANE-Firebase/wiki/Database---Read-Data-and-Change-Events

https://github.com/distriqt/ANE-Firebase/wiki/Database---Lists

**Comment by Shaunmax** [comment]
No, I want the query option which returns the record in ascending order. i
wanted to know if its possible to detect end_of_query results, when the
result doesn't meet requested limit, eg i need the result of Top 10 players
but there are only 5 participants data available in the database.

**Comment by marchbold** [comment]
Try using the `DatabaseReferenceEvent.VALUE_CHANGED` event then instead of the `CHILD_ADDED`.

**Comment by Shaunmax** [comment]
When using VALUE_CHANGED its not returning any data.

**Comment by marchbold** [comment]
This will only be dispatched when you first create the query or when it changes. So it's important you are using a single instance in this case.

**Comment by Shaunmax** [comment]
Yes, that how I implemented it, but still no result! I am querying
yesterday result only once, when the app starts:-

```
yesterdayQuery =
FirebaseDatabase.service.getReference("myDatabase").child("leaderboard").child(yesterday_server).orderByChild("gScore").limitToFirst(10);
yesterdayQuery.addEventListener(DatabaseReferenceChildEvent.CHILD_CHANGED,
handleYesterdayList);
```

**Comment by marchbold** [comment]
Hi, 

That's the wrong event, you need to add the `DatabaseReferenceEvent.VALUE_CHANGED` event

**Comment by Shaunmax** [comment]
When I changed it, I got this error:-

TypeError: Error #1034: Type Coercion failed: cannot convert com.distriqt.extension.firebase.database.events::DatabaseReferenceEvent@cb0cb5f1 to com.distriqt.extension.firebase.database.events.DatabaseReferenceChildEvent.
	at flash.events::EventDispatcher/dispatchEventFunction()
	at flash.events::EventDispatcher/dispatchEvent()
	at _Query/extContext_statusHandler()

**Comment by Shaunmax** [comment]
my bad i forgot to update the listener function

**Comment by Shaunmax** [comment]
The issue is solved with DatabaseReferenceEvent.VALUE_CHANGED.

**Comment by Shaunmax** [comment]
I have another doubt regarding database caching? even after using 

> keepSynced(true);

most of the time, the value I get is outdated or old. How can I make sure the cached version is deleted or removed!

**Comment by marchbold** [comment]
The caching is internal to the Firebase SDK so we don't control how it's implemented. 

If you are having issues with it perhaps read through the Firebase Realtime Database documentation to see best approaches.

**Comment by Shaunmax** [comment]
Ok, thanks

**Comment by Shaunmax** [comment]
Hi Michael,

I just realised now that, when I query with DatabaseReferenceEvent.VALUE_CHANGED, its always returning the same set of values no matter what. Its like picking the value from cache.

I found this after pushing the build live, so when a playerA loads the app, the query fetches first 10 players data and only updates it when the playerA gets a new high score. If another playerB gets a new high score and playerA fetches the query, playerB's score wont be there. That's strange, could please tell me why this behaviour, is it fetching from the cache?
